pkg/aspect/quotasManager.go

```diff
@@ -58,94 +74,129 @@ func (m *quotasManager) NewAspect(c *config.Combined, a adapter.Builder, env ada
 		},
 	}
 
-	defs := make(map[string]*adapter.QuotaDefinition)
+	metadata := make(map[string]*quotaInfo, len(desc))
+	defs := make(map[string]*adapter.QuotaDefinition, len(desc))
 	for _, d := range desc {
-		dur, _ := ptypes.DurationFromProto(d.Expiration)
+		quota := findQuota(params.Quotas, d.Name)
+		if quota == nil {
+			env.Logger().Warningf("No quota found for descriptor %s, skipping it", d.Name)
+			continue
+		}
 
-		defs[d.Name] = &adapter.QuotaDefinition{
-			MaxAmount: d.MaxAmount,
-			Window:    dur,
+		// TODO: once we plumb descriptors into the validation, remove this err: no descriptor should make it through validation
+		// if it cannot be converted into a QuotaDefinition, so we should never have to handle the error case.
+		def, err := quotaDefinitionFromProto(&d)
+		if err != nil {
+			_ = env.Logger().Errorf("Failed to convert quota descriptor '%s' to definition with err: %s; skipping it.", d.Name, err)
+			continue
+		}
+
+		defs[d.Name] = def
+		metadata[d.Name] = &quotaInfo{
+			labels:     quota.Labels,
+			definition: def,
 		}
 	}
 
-	aspect, err := a.(adapter.QuotasBuilder).NewQuotasAspect(env, c.Builder.Params.(adapter.AspectConfig), defs)
+	asp, err := a.(adapter.QuotasBuilder).NewQuotasAspect(env, c.Builder.Params.(adapter.AspectConfig), defs)
 	if err != nil {
 		return nil, err
 	}
 
 	return &quotasWrapper{
-		manager:     m,
-		descriptors: desc,
-		inputs:      c.Aspect.GetInputs(),
-		aspect:      aspect,
+		manager:  m,
+		metadata: metadata,
+		aspect:   asp,
 	}, nil
 }
 
 func (*quotasManager) Kind() Kind                                                     { return QuotasKind }
 func (*quotasManager) DefaultConfig() adapter.AspectConfig                            { return &aconfig.QuotasParams{} }
 func (*quotasManager) ValidateConfig(adapter.AspectConfig) (ce *adapter.ConfigErrors) { return }
 
-func (w *quotasWrapper) Execute(attrs attribute.Bag, mapper expr.Evaluator) (*Output, error) {
-	// labels holds the generated attributes from mapper
-	labels := make(map[string]interface{})
-	for attr, e := range w.inputs {
-		if val, err := mapper.Eval(e, attrs); err == nil {
-			labels[attr] = val
+func (w *quotasWrapper) Execute(attrs attribute.Bag, mapper expr.Evaluator, ma APIMethodArgs) (*Output, error) {
+	qma, ok := ma.(*QuotaMethodArgs)
+
+	// TODO: this conditional is only necessary because we currently perform quota
+	// checking via the Check API, which doesn't generate a QuotaMethodArgs
+	if !ok {
+		qma = &QuotaMethodArgs{
+			Quota:           "RequestCount",
+			Amount:          1,
+			DeduplicationID: strconv.FormatInt(atomic.AddInt64(&w.manager.dedupCounter, 1), 16),
+			BestEffort:      false,
 		}
 	}
 
-	// TODO: for now we don't support dedup semantics
-	dedupID := strconv.FormatInt(atomic.AddInt64(&w.manager.dedupCounter, 1), 16)
+	info, ok := w.metadata[qma.Quota]
+	if !ok {
+		return &Output{Code: code.Code_INVALID_ARGUMENT}, fmt.Errorf("unknown quota '%s'", qma.Quota)
```

Comment:
Is this correct? We should be using gogoprotos which should generate the enum `Code.INVALID_ARGUMENT`